src/sstruct_mv/sstruct_matrix.c

```diff
@@ -1082,33 +1112,80 @@ hypre_SStructUMatrixSetBoxValues( hypre_SStructMatrix *matrix,
 
                start = hypre_BoxIMin(int_box);
                hypre_BoxGetSize(int_box, loop_size);
-               /*FIXME: Currently works only for the default boxloop (see GetIndex below) */
-               zypre_BoxLoop2Begin(ndim, loop_size,
-                                   box,  start, stride, mi,
-                                   value_box, start, stride, vi);
-#ifdef HYPRE_USING_OPENMP
-#pragma omp parallel for private(HYPRE_BOX_PRIVATE) HYPRE_SMP_SCHEDULE
+
+#if defined(HYPRE_USING_CUDA) || defined(HYPRE_USING_DEVICE_OPENMP)
+               hypre_assert(ndim <= 3);
```

Comment:
@liruipeng Can we make this at least 4D?  Are there other 3D limitations in the GPU code?

Reply:
@rfalgout There are a lot more we need to do to address 4D. GPU BoxLoops do not support 4D yet. Should we make the effort in another branch later? Thanks!

Reply:
Okay, since it only affects GPUs.  Thanks!